### ISSUE TITLE
Backfill: ~1,210 approved examples were scored against a placeholder checklist

### ISSUE THREAD

**Issue by kreuzhofer** [kreuzhofer/chat3d-app#34]
Follow-up to #33, which fixed the defect but deliberately scoped the data out.

## The situation

Until #33, the visual judge received a checklist of literal `undefined` placeholders in 58% of evaluations (1,873 of 3,228 stored prompts). Every affected example's `visual_score`, `eval_score` and `approval_status` derives from that evaluation.

**1,210 of them are auto-approved**, which means they are in the fine-tuning training set.

| cohort | examples | approved | avg visual |
| --- | --- | --- | --- |
| placeholder checklist | 1,333 | 1,210 (91%) | 8.48 |
| real or no checklist | 1,257 | 1,216 (97%) | 8.67 |

## The cost is lower than it looks

The source data was never lost. Of 2,855 prompt records, **2,485 still hold both** the plain `verification_checklist` and the annotated `verification_criteria`, and the plain checklist is already phrased as visual questions ("Does the model have exactly two arms meeting at a single elbow junction?").

So this is a **re-run of visual evaluation only** — roughly one VLM call per example. No spec regeneration, no codegen, no re-render.

## Do not assume the direction

The obvious hypothesis is "no checklist made the judge lenient, so scores will drop". The data does not support it. Controlling for category, only four have enough of both cohorts to compare:

| category | placeholder | real | delta |
| --- | --- | --- | --- |
| bd_warehouse | 8.41 (n=70) | 8.33 (n=6) | −0.08 |
| Electronic Components | 8.49 (n=101) | 7.60 (n=5) | **−0.89** |
| Generic Enclosures | 7.83 (n=92) | 7.00 (n=8) | **−0.83** |
| Missing Examples | 8.55 (n=116) | 8.86 (n=993) | **+0.31** |

Two of the four have the wrong sign, three have samples too small to conclude anything, and the only comparison with statistical weight goes the other way. Any plan that pre-commits to an outcome is guessing.

## Proposed sequencing

1. **Re-evaluate a stratified sample of ~100**, weighted toward the categories where the real-checklist rate was 0–1% (Gridfinity Storage, Hinges, PCB Cases, Generic Enclosures) — that is where a working checklist should bite hardest.
2. **Measure the delta** against the stored scores: mean change, and how many examples cross the auto-approval threshold in each direction. The threshold crossings matter more than the mean, because they are what changes the dataset.
3. **Decide the full backfill from that number.** If the delta is negligible, provenance-marking is enough. If it is material, re-run the affected set.

## Provenance is needed either way

Whatever is decided, examples scored before and after #33 are not comparable, and anything treating this corpus as a reference set — judge benchmarking, a gold set, fine-tuning data — needs to know which side of the fix each row is on. A mixed corpus teaches two different standards.

The cheapest marker is already in the data: `vlm_system_prompt LIKE '%. undefined%'` identifies the affected rows exactly. A durable flag or a re-eval timestamp would be better than re-deriving it by string match forever.

## Acceptance criteria

- [ ] A stratified sample of affected examples is re-evaluated against the fixed pipeline, without mutating the originals until the result is reviewed.
- [ ] The delta is reported as: mean score change, and count of examples crossing the auto-approval threshold in each direction, broken down by category.
- [ ] A decision on the full backfill is recorded on this issue, with the number that drove it.
- [ ] Affected rows are distinguishable from unaffected ones by something more durable than a `LIKE` on the stored prompt.
- [ ] If a full backfill runs, approval-rate changes per category are reported before and after.

## Out of scope

- The code fix itself (#33, done).
- Changing the rubric, the judge model, or the auto-approval threshold — a backfill must re-measure with the *current* configuration, or it confounds two changes.
- Deleting or re-generating examples. This is re-evaluation of existing renders only.

**Comment by kreuzhofer** [comment]
## Precondition: the #33 fix is committed but NOT deployed

As of 2026-09-01, `8ecb8e8` is pushed but the `chat3d-backend` container still runs the pre-fix image. A deploy was started and cancelled.

**Re-evaluating anything before deploying reproduces the bug and measures nothing.** This is the one step that invalidates the whole exercise, so do it first:

```
docker compose build backend && docker compose up -d backend
# confirm the fix is in the image:
docker exec chat3d-backend ls /app/src/utils/verification-criteria.ts
```

Note `docker compose up -d <service>` recreates any service whose running container is stale against a newer image, not only the one named — check nothing is mid-run before deploying.

## Working facts for the sample

- Affected rows are identified exactly by `vlm_system_prompt LIKE '%. undefined%'` — 1,333 production examples, 1,210 of them auto-approved.
- Source data was never lost: 2,485 of 2,855 `workbench_example_prompts` rows still hold both `verification_checklist` (plain, already phrased as visual questions) and `verification_criteria` (annotated). Re-evaluation is a visual-eval re-run, not a pipeline pass.
- Stratify toward the categories that had 0–1% real checklists: Gridfinity Storage, Hinges, PCB Cases, Generic Enclosures.
- Report auto-approval threshold crossings per category, not only mean delta.
- Do not pre-commit to a direction — see the table in the issue body; two of four comparable categories moved the wrong way.

## One confound to hold fixed

The `vlm_eval` purpose currently maps to the local `qwen38-nvfp4-spark`, not a frontier model. Measured over the same prompt population, that judge showed a suspected false-negative rate of 9.09% (6/66) against Sonnet's 0.42% (11/2,618) — a ~22× difference — so it is not a neutral instrument.

Whatever judge is used for the sample must be the same one used for the stored scores being compared against, or the delta measures two changes at once. The stored scores in question were produced by Sonnet.

**Comment by kreuzhofer** [comment]
## Result: the delta is material. Recommending the full backfill.

**Decision driver: ~160 of the 1,210 approved examples (13.2%) would lose auto-approval.** The mean visual score moves −0.03, which is nothing. The two numbers disagree because the mean is the wrong instrument here, exactly as the issue anticipated.

### What was run

125 affected examples (`eval_checklist_state = 'placeholder'`, production only), re-evaluated against their existing renders with the real checklist through the post-#33 `deriveVisualChecklist` path. 125/125 completed, 0 errors. Stored as VLM experiment `96f4f702-4346-4103-8555-4b816023412f`; **no original row was mutated**.

Stratified per the plan, weighted toward the 0–1% real-checklist categories, plus 25 spread over five large high-scoring categories so the ~1,050 approved rows living there were not left unmeasured.

**Judge.** All 1,333 stored scores in this cohort were produced by Sonnet 4.6, so the re-run had to be Sonnet. Bedrock credentials are dead (last successful call 2026-06-18), so it ran through the Anthropic provider on `claude-sonnet-4-6`, same $3/$15, **thinking off** — matched to the stored condition, which was verified rather than assumed: 0 of 22,623 stored Sonnet vlm_evaluation events carry reasoning tokens, and the instrument works, since Bedrock spec_generation reports them on 3,545 events. Added as a new model row; `llm_purpose_map` was not touched.

Composite and approval were recomputed with the production helpers at current settings (threshold 7.5, code weight 0.4), holding the code score and weights at their stored values, so the visual score is the only thing that moved.

### Per category

| category | n | visual was → now | Δ | approved was → now | +/− |
| --- | ---: | --- | ---: | --- | --- |
| Boolean Operations | 5 | 9.00 → 9.00 | +0.00 | 5 → 5 | +0 / −0 |
| Extrusions and Revolutions | 5 | 8.80 → 8.80 | +0.00 | 5 → 4 | +0 / −1 |
| Generic Enclosures | 25 | 7.88 → 7.80 | −0.08 | 19 → 15 | +1 / −5 |
| Gridfinity Storage | 19 | 5.74 → 5.47 | −0.26 | 4 → 1 | +1 / −4 |
| Hinges | 24 | 5.29 → 5.29 | +0.00 | 5 → 4 | +1 / −2 |
| Missing Examples | 5 | 8.60 → 8.20 | −0.40 | 5 → 4 | +0 / −1 |
| PCB Cases | 32 | 5.63 → 5.78 | **+0.16** | 1 → 2 | +1 / −0 |
| Primitives | 5 | 9.20 → 9.20 | +0.00 | 5 → 4 | +0 / −1 |
| Sketch Operations | 5 | 9.00 → 9.00 | +0.00 | 5 → 5 | +0 / −0 |
| **ALL** | **125** | **6.69 → 6.66** | **−0.03** | **54 → 44** | **+4 / −14** |

### The direction warning in the issue was right

PCB Cases — one of the categories predicted to suffer most — moved **up** (+0.16, and gained an approval). Scores move both ways: a placeholder checklist gave the judge nothing concrete to verify, so it scored conservatively. A 12V battery-holder enclosure went 6.0 → 8.0 with 6/6 real checklist items passing.

And the effect is **not confined to the hard categories**. Three of the five big high-scoring categories each lost an approval from a sample of 5 — 3/25 = 12%. Those categories hold most of the 1,210.

### Threshold crossings, split by cause

14 lost approval, 4 gained, **net −10 of 54 (18.5%)**. The 14 are not one phenomenon:

- **8 failed a real checklist question while still scoring ≥ 7.5.** These are gated by the checklist itself — questions a placeholder list could never have asked. Directly attributable to the #33 fix. Example: a wall switch plate scoring 8.0 with 0/1 checklist items passing.
- **6 fell below 7.5 on score alone.** This single-arm design cannot separate those from judge run-to-run variance.

### Extrapolation, with its assumptions stated

The sample over-weights hard categories, so the raw 18.5% cannot be applied to the corpus. Reweighting by each category's actual approved count, using the measured rate where the sample supports it and the pooled big-category rate (12.0%) elsewhere:

| basis | approved | rate |
| --- | ---: | ---: |
| Generic Enclosures (measured) | 70 | 26.3% |
| Hinges (measured) | 5 | 40.0% |
| Gridfinity Storage (measured) | 4 | 100.0% |
| all large categories (pooled / assumed) | 1,131 | 12.0% |

**≈160 of 1,210 lose approval (13.2%); ≈92 attributable to the checklist gate** if the measured 8/14 split holds.

Six categories totalling 540 approved rows (Arrays and Patterns, Electronic Components, Surface Modifications, Simple Everyday Objects, Mechanical Components, bd_warehouse) were **not sampled** and carry the pooled 12% assumption. They are the largest source of uncertainty in the 160.

### Recommendation

**Run the full backfill on all 1,333.** ~160 examples sitting in the fine-tuning set with approval they would not now receive is worth ~1,333 VLM calls (≈$30, ≈5h sequential at the observed ~15s each). Provenance-marking alone is not enough at this magnitude.

Caveats to carry into that run:
- Hold the judge fixed (Sonnet 4.6, thinking off) or the backfill re-confounds what this measured.
- Without a control arm, ~43% of the crossings (6/14) are not separable from variance. If that matters for the training-set decision, a placeholder-replay arm on the same 125 would settle it for ~$3.
- Code review also lost its code-only criteria emphasis to #33 (the filter `visibility === "code"` matched nothing on the broken shape). It received no garbage — its prompts are clean — but it is a second, unmeasured degradation. Out of scope here.

### Acceptance criteria status

- [x] Stratified sample re-evaluated against the fixed pipeline, originals unmutated
- [x] Delta reported as mean change + threshold crossings per category
- [x] Decision recorded with the number that drove it (~160 / 13.2%)
- [x] Affected rows durably distinguishable — `workbench_examples.eval_checklist_state` (`real` | `empty` | `placeholder`), backfilled to 1,881 placeholder rows (1,333 production, 1,210 approved), **0 disagreements** with the old `LIKE`
- [ ] Full-backfill approval-rate changes per category — pending the decision above

### One blocker worth recording

The harness could not have measured this before today: the VLM experiment executor never passed a checklist at all, so it would have compared placeholder against *nothing* and produced a plausible number answering a different question. Fixed alongside this work.

**Comment by kreuzhofer** [comment]
## Full backfill complete — approval rate 90.8% → 81.0%

1,333/1,333 re-evaluated, **0 failures**, 5h03m, **$43.90** (1,333 calls, $0.033 avg, 13.6s avg). Visual evaluation only; code score, assertion pass rate and weights held at stored values. Rollback snapshot of every pre-run row is in `workbench_examples_pre_issue34`.

**144 examples lost auto-approval, 14 gained — net −130.** Predicted from the sample was ~160 (13.2%); actual 144 (**11.9%**).

### Before → after, by category

| category | n | approved before | after | lost | gained | rate before → after | visual before → after |
| --- | ---: | ---: | ---: | ---: | ---: | --- | --- |
| Boolean Operations | 138 | 132 | 109 | 24 | 1 | 95.7% → 79.0% | 8.61 → 8.30 |
| Generic Enclosures | 92 | 70 | 51 | 21 | 2 | 76.1% → 55.4% | 7.83 → 7.66 |
| Surface Modifications | 98 | 95 | 77 | 19 | 1 | 96.9% → 78.6% | 8.71 → 8.36 |
| Extrusions and Revolutions | 127 | 122 | 110 | 13 | 1 | 96.1% → 86.6% | 8.72 → 8.45 |
| Arrays and Patterns | 101 | 100 | 89 | 11 | 0 | 99.0% → 88.1% | 8.76 → 8.55 |
| Missing Examples | 116 | 106 | 99 | 9 | 2 | 91.4% → 85.3% | 8.55 → 8.61 |
| Sketch Operations | 105 | 105 | 96 | 9 | 0 | **100.0% → 91.4%** | 8.81 → 8.61 |
| Mechanical Components | 90 | 88 | 80 | 8 | 0 | 97.8% → 88.9% | 8.61 → 8.53 |
| bd_warehouse Examples | 70 | 64 | 57 | 8 | 1 | 91.4% → 81.4% | 8.41 → 8.07 |
| Simple Everyday Objects | 95 | 93 | 88 | 7 | 2 | 97.9% → 92.6% | 8.60 → 8.54 |
| Electronic Components | 101 | 100 | 95 | 5 | 0 | 99.0% → 94.1% | 8.49 → 8.52 |
| Primitives | 125 | 125 | 121 | 4 | 0 | **100.0% → 96.8%** | 9.41 → 9.39 |
| Gridfinity Storage | 19 | 4 | 2 | 3 | 1 | 21.1% → 10.5% | 5.74 → 5.53 |
| Hinges | 24 | 5 | 4 | 2 | 1 | 20.8% → 16.7% | 5.29 → 5.25 |
| PCB Cases | 32 | 1 | 2 | 1 | 2 | **3.1% → 6.3%** | 5.63 → 5.75 |
| **TOTAL** | **1,333** | **1,210** | **1,080** | **144** | **14** | **90.8% → 81.0%** | |

### The issue's original targeting was backwards

The plan stratified toward Gridfinity Storage, Hinges and PCB Cases, on the reasoning that a working checklist should bite hardest where the real-checklist rate was 0–1%. It bit *least* there — those three lost 6 approvals between them, because they had almost none to lose (4, 5 and 1 approved respectively). PCB Cases came out **net positive**.

The damage was concentrated in the categories that looked healthiest. Boolean Operations, Surface Modifications and Generic Enclosures account for **64 of the 144 losses**. Two categories sat at a perfect 100% approval rate and no longer do — Sketch Operations 100% → 91.4%, Primitives 100% → 96.8%.

That is the finding worth keeping: **the placeholder checklist was inflating approval precisely where the corpus looked strongest**, and a 100% approval rate was evidence of a broken measurement rather than a healthy category. The hard categories were already being rejected on general impression, so a real checklist changed little there.

### Acceptance criteria

- [x] Stratified sample re-evaluated against the fixed pipeline, originals unmutated
- [x] Delta reported as mean change + threshold crossings per category
- [x] Decision recorded with the number that drove it (~160 predicted; 144 actual)
- [x] Affected rows durably distinguishable — `eval_checklist_state`, now `real`/`empty` for all 1,333
- [x] Full backfill approval-rate changes per category reported before and after

### Carry-forward, not addressed here

1. **`vlm_eval` still maps to the local `qwen38-nvfp4-spark`.** This corpus is now Sonnet-judged against real checklists; every new evaluation is judged by a model measured at a 9.09% suspected false-negative rate against Sonnet's 0.42%. The corpus starts re-mixing standards with the next generation run.
2. **The composites written here embed #43** (adaptive weight shift applied twice). Deliberate — it keeps the backfill consistent with the sample the decision rested on — and correctable from stored scores with no further LLM calls.
3. **235 production examples had no checklist at all.** Not corrupted, so out of scope, but the judge had nothing concrete to verify there either.
4. **Code review's code-only criteria emphasis was also lost to #33** (empty list, not garbage). Unmeasured.

Leaving this open until the branch carrying the fixes is merged.

**Comment by kreuzhofer** [comment]
## The missing control arm, obtained for free — the effect holds

The judge head-to-head (experiment `7337a398`) re-runs Sonnet 4.6 over the **same 125 examples** with the **same real checklists** and the **same config**. Since those rows now carry the backfilled Sonnet scores, that run is a straight repeatability test: same judge, same input, nothing changed. It is exactly the control arm this measurement was run without.

### Sonnet's own noise floor

| | |
| --- | ---: |
| mean delta | −0.080 |
| mean absolute delta | **0.288** |
| identical score | 75.2% |
| within ±1 | 96.0% |
| max swing | 2.0 |

**The judge is not deterministic.** A quarter of scores move on a plain re-run, by 0.29 on average.

### Which settles the mean-vs-crossings question

The sample's headline mean delta was **−0.03**. Sonnet's own run-to-run mean delta is **−0.08** — nearly three times larger. The mean delta this exercise was nearly reported on was **smaller than the instrument's noise**. Anyone reading that −0.03 as "no effect" would have been reading noise.

### And confirms the crossings were real

| | lost | gained | net |
| --- | ---: | ---: | ---: |
| pure variance (Sonnet → Sonnet) | 3 | 3 | **0** |
| checklist fix (placeholder → real) | 14 | 4 | **−10** |

Variance churns **symmetrically and cancels**. The checklist fix produced a **one-directional** shift roughly 3× the noise band in magnitude, and with none of the cancellation. Netting to zero is what a null result looks like here, and that is not what the fix produced.

Reconciling with the earlier split: of the 14 losses, 8 were checklist-gated (scored ≥7.5 but failed a real question) and 6 were score-driven. The control suggests ~3 of those 6 are attributable to noise, leaving ~11 of 14 real — consistent with the 8 that were independently attributable.

**The decision to run the full backfill stands, now with a measured control rather than an assumption.** The caveat recorded earlier — "~43% of the crossings are not separable from variance" — is resolved: most of them are separable, and the ones that are not are outnumbered by the ones that are.

### Carry-forward

This bears directly on #36 (*the visual judge is not a stable instrument*). A 0.288 mean absolute self-disagreement is the **floor** on any single-run comparison of judges or examples. Any future measurement whose effect is smaller than that needs repeat runs, not a single pass — including per-example gold-set labelling, where one Sonnet run is now known to be a noisy label.
